### DISCUSSION TITLE
No prompt to clone git repo

### DISCUSSION THREAD

**Discussion by indeedwatson** [zeapo/Android-Password-Store#1317] [Q&A] [answered]
Coming from this: https://github.com/android-password-store/Android-Password-Store/issues/1118#issuecomment-777971632

I apologize that it wasn't the right place, I wasn't even aware of this Discussions section, is this a new github feature?

But in any case, in that thread:

> When you first launch the app, select the "Clone Git repo" option which will lead you to the clone configuration screen. Enter your repository URL, pick how you want to authenticate and hit the clone button.

I can't "first launch the app" anymore since I already launched it before, but when I do open it, there is no option to clone a git repo, anywhere. There is a + button to add a folder or password, and there's settings, but nowhere in the settings do I see a place to add a git repo.

**Comment by Skrilltrax** [comment, marked as answer]
You can go to settings and delete your current repository, that will launch the first run flow again

**Reply by indeedwatson** [reply]
There's no such options in settings, these are all the options:

- enable autofill
- password fill organization
- default username
- custom domains
- generate ssh key pair
- external repository (checked)
- select external repository (none selected)
- password generator type
- app theme
- password sort order
- recursive filtering
- open search on start
- show hidden folders
- enable biometric authentication
- password copy timeout
- show the password
- show extra content
- automatically copy password
- export passwords
- clear clipboard 20 times
- debug logging
- version

That's it.

**Reply by msfjarvis** [reply]
Uncheck 'External repository' and the option to delete will pop back up. We really need to improve our reliability in the Git space...

**Reply by indeedwatson** [reply]
Okay that worked, and I managed to solve another issue I had with openkeychain where I couldn't add the key.

But now I'm getting authentication issue and I'm not sure what to do at this point. Is https with password discouraged?

**Reply by msfjarvis** [reply]
> Okay that worked, and I managed to solve another issue I had with openkeychain where I couldn't add the key.
> 
> But now I'm getting authentication issue and I'm not sure what to do at this point. Is https with password discouraged?

It's not, did you provide a username in the URL? The typical format is `https://username@domain.tld/path/to/repo.git`. If you're using GitHub, then HTTPS with your account password is now discouraged and using a personal access token as a password is the way to go.

**Reply by indeedwatson** [reply]
I was trying with the ssh url format.

Trying the https format with openkeychain, I get `Authentication is required but no CredentialsProvider has been registered`, which seems to be cause it's a private repo.

I'm using gitlab.

**Reply by indeedwatson** [reply]
Turns out I didn't have pgp set up with gitlab... I was sure I did. But in any case, I just added it, but I keep getting the Auth fail.

Do I need to "refresh" the key on OpenKeyChain or something?

**Reply by msfjarvis** [reply]
OpenKeychain only works for SSH connections, because it essentially generates an SSH key from your PGP key.

First, pick what protocol you wish to use: SSH or HTTPS.

For SSH, go to OpenKeychain, select your key, click on the "Advanced" menu item from the three dots at the top, then "Share as SSH public key" from the Share tab. This will give you an SSH key that should then be added to your GitLab repository (if the concept of per-repository deploy keys exists on the platform) or to your GitLab account.

For HTTPS, generate a personal access token that can access private repositories, set the URL in the format previously mentioned and provide the token as a password when prompted

Hope this helps.

**Reply by indeedwatson** [reply]
Thank you for the detailed explanation. I followed the steps and I finally managed to get it to work, I'll have to make a text file for future reference.

I wonder, how is a user supposed to know this? There's an SSH option already, so how am I meant to know that OpenKeyChain option also needs a SSH (not to mention all the rest of the steps)?

And what is the difference between the SSH option in the menu, and using SSH with OpenKeyChain?

And yet another question, when I try to use the SSH key option, it never prompts me to select it, it just says "No provider available for Unknown key file". And again I wonder, how is a user meant to know what to do from there?

I hope all these questions aren't annoying, I'm trying to formulate them in a way that exposes whether it's a me problem, or if there can be more intuitive ways to communicate to clueless users what to do.


Thanks for all the work you do!

**Reply by msfjarvis** [reply]
> Thank you for the detailed explanation. I wonder, how is a user supposed to know this? There's an SSH option already, so how am I meant to know that OpenKeyChain option also needs a SSH (not to mention all the rest of the steps)?

The `pass` ecosystem that this app ties into is a super Linux-y one, so everything was developed with the assumption that every APS user is also a seasoned Linux user and knows their way around the utter mess that is GnuPG. We've been hard at work dismantling those assumptions and things are now better, but still far from ideal. The gaping hole that is our lacklustre documentation needs a serious time investment which we have been unable to make due to various reasons. I'm hoping to push out some of what we're discussing here into the docs site over the weekend.

> 
> And what is the difference between the SSH option in the menu, and using SSH with OpenKeyChain?
> 

With the SSH key option, the SSH key is not derived from a GnuPG key but generated (or imported if you already have one) and optionally stored in your device's secure enclave (if it has one).

> And yet another question, when I try to use the SSH key option, it never prompts me to select it, it just says "No provider available for Unknown key file". And again I wonder, how is a user meant to know what to do from there?
> 

This is very likely a bug. Unfortunately there are too many ways for users to trip up and cause the app to enter a bad state (another problem we're actively tackling), so clearing app data and going through the first-time setup process should offer you the option to generate or import the SSH key when you select SSH as your authentication method.

> I hope all these questions aren't annoying, I'm trying to formulate them in a way that exposes whether it's a me problem, or if there can be more intuitive ways to communicate to clueless users what to do.
> 

Not at all, I welcome the queries. Part of the documentation problem is that it's hard for us as maintainers to put ourselves in our users' place and figure out the pain points that need documenting and fixes, so these interactions help us figure out exactly what users are struggling with so we can make the experience more robust and straightforward.

**Reply by indeedwatson** [reply]
> The pass ecosystem that this app ties into is a super Linux-y one, so everything was developed with the assumption that every APS user is also a seasoned Linux user and knows their way around the utter mess that is GnuPG.

Well that is a fair assumption, but what I'm experiencing is that when using tools within the linux environment (even when they're obscure) there's more of a handrail, so to speak, to navigate these tools than there is in mobile/android environments. Direct interaction like this certainly helps tho. The instructions you gave me, perhaps with some screenshots, would be great for documentation imo.
